mempool/mempool.go

```diff
@@ -140,3 +144,516 @@ func PostCheckMaxGas(maxGas int64) PostCheckFunc {
 		return nil
 	}
 }
+
+//--------------------------------------------------------------------------------
+
+var _ Mempool = &basemempool{}
+
+type basemempool struct {
+	mempoolImpl
+
+	// Atomic integers
+	height   int64 // the last block Update()'d to
+	txsBytes int64 // total size of mempool, in bytes
+
+	// Notify listeners (ie. consensus) when txs are available
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // fires once for each height, when the mempool is not empty
+
+	// Keep a cache of already-seen txs.
+	// This reduces the pressure on the proxyApp.
+	cache     txCache
+	preCheck  PreCheckFunc
+	postCheck PostCheckFunc
+
+	config *cfg.MempoolConfig
+
+	// Exclusive mutex for Update method to prevent concurrent execution of
+	// CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.
+	updateMtx sync.RWMutex
+
+	wal          *auto.AutoFile // a log of mempool txs
+	proxyAppConn proxy.AppConnMempool
+
+	logger  log.Logger
+	metrics *Metrics
+}
+
+type mempoolImpl interface {
+	Size() int
+	addTx(*mempoolTx, uint64)
```

Comment:
also add a comment explaining all private methods will assume locks are held by `basemempool` so the impl themselves won't need to